### PR TITLE
Cleanup CI (#5040)

### LINKED ISSUE #5040: CI Failing with Out of Disk
**Describe the bug**

CI is failing due to running out of disk space https://github.com/apache/arrow-datafusion/actions/runs/3995952868/jobs/6855600796

This correlates with enabling the dyn_arith_dict feature in CI - https://github.com/apache/arrow-datafusion/pull/4999

**To Reproduce**

**Expected behavior**

We should not exceed the disk capacity of the runners, either by splitting up the CI into smaller steps and cleaning the target directory inbetween, or splitting the CI into multiple smaller workflows like we do for arrow-rs

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #5040

# Rationale for this change

* Splits out doctests into separate CI job (as they result in a lot of generated code)
* Splits out examples into separate CI job (as they take a while to compile and run)
* Simplifies the hash collision and pythin CI jobs
* Replaces cancel workflow with [concurrency groups](https://docs.github.com/en/actions/using-jobs/using-concurrency#example-only-cancel-in-progress-jobs-or-runs-for-the-current-workflow)
* Miscellaneous cleanups

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -32,10 +36,6 @@ jobs:
     runs-on: ubuntu-latest
     container:
       image: amd64/rust
-      env:
```
Comment (tustvold) [inline .github/workflows/rust.yml:35]: This has been moved into setup-builder

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -478,19 +485,14 @@ jobs:
       - name: Run tests
         run: |
           cd datafusion
-          # Force all hash values to collide
-          cargo test --all --features=force_hash_collisions
+          cargo test --lib --tests --features=force_hash_collisions
```
Comment (tustvold) [inline .github/workflows/rust.yml:488]: I couldn't see a compelling reason to also run the doctests here

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -356,10 +374,9 @@ jobs:
         uses: ./.github/actions/setup-builder
         with:
           rust-version: stable
-      - name: Run tests
-        run: |
-          cd datafusion
-          cargo test --features=pyarrow
+      - name: Run datafusion-common tests
+        run: cargo test -p datafusion-common --features=pyarrow
```
Comment (tustvold) [inline .github/workflows/rust.yml:378]: The only pyarrow code is in datafusion-common AFAICT

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- this looks great

I also manually looked at the output of the test CI jobs and ensured all the tests ran the way I expected.

**Comment by alamb** [comment]
I am going to merge this in so we can unblock other PRs -- we can handle comments / issues as follow on.

Thanks again @tustvold

**Pull request merged**